rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1197,7 +1277,119 @@ impl SystemState {
+                stop_contexts,
+            } => {
+                let stop_contexts = std::mem::take(stop_contexts);
+                self.status = CanisterStatus::Running {
+                    call_context_manager: std::mem::take(call_context_manager),
+                };
+                stop_contexts
+            }
+
+            CanisterStatus::Stopped => {
+                self.status = CanisterStatus::new_running();
+                Vec::new()
+            }
+        }
+    }
+
+    /// Transitions the canister into `Stopping` state.
+    ///
+    /// If the canister was `Running` or `Stopping`, remembers the stop context, so
+    /// that it can be responded to once the canister has fully stopped. If the
+    /// canister was already `Stopped`, returns the stop context.
+    pub fn begin_stopping(
+        &mut self,
+        stop_context: StopCanisterContext,
+    ) -> Option<StopCanisterContext> {
+        match &mut self.status {
+            // Return the stop context, nothing to do here.
+            CanisterStatus::Stopped => Some(stop_context),
+
+            CanisterStatus::Stopping {
+                ref mut stop_contexts,
+                ..
+            } => {
+                // Add the message so we can respond to it once the canister has fully stopped.
+                stop_contexts.push(stop_context);
+                None
+            }
+
+            CanisterStatus::Running {
+                call_context_manager,
+            } => {
+                // Transition the canister into the stopping state.
+                self.status = CanisterStatus::Stopping {
+                    call_context_manager: std::mem::take(call_context_manager),
+                    // Track the stop message to respond to it once the canister is fully stopped.
+                    stop_contexts: vec![stop_context],
+                };
+                None
+            }
+        }
+    }
+
+    /// Tries to transition a `Stopping` canister into the `Stopped` state. No-op if
+    /// the canister is `Running` or already `Stopped`.
+    ///
+    /// Returns a tuple of:
+    ///  * a boolean indicating whether the canister has stopped,
+    ///  * all stop contexts if the canister has stopped; or the expired stop
+    ///    contexts only if the canister is still stopping.
+    #[must_use]
+    pub fn try_stop_canister(
+        &mut self,
+        is_expired: impl Fn(&StopCanisterContext) -> bool,
+    ) -> (bool, Vec<StopCanisterContext>) {
+        match self.status {
+            // Canister is not stopping so we can skip it.
+            CanisterStatus::Running { .. } | CanisterStatus::Stopped => (false, Vec::new()),
+
+            // Canister is ready to stop.
+            CanisterStatus::Stopping {
+                ref call_context_manager,
+                ref mut stop_contexts,
+            } if call_context_manager.callbacks().is_empty()
+                && call_context_manager.call_contexts().is_empty() =>
```

Comment:
"Call context manager is empty" doesn't sound right. And a "has no call contexts or callbacks" method is almost as verbose, but slightly more opaque.

Technically, it should be sufficient to check that `call_contexts().is_empty()`, since you cannot have any callbacks without call contexts, but that also may not be immediately obvious to someone reading the code.

Regarding `unexpired_callbacks()`, they are supposed to be a subset of the best-effort callbacks in `callbacks`. And we already have a (soft) invariant check in `stats_ok()` for precisely this.